ADR/0015-integration-service-two-phase-architecture.md

```diff
@@ -0,0 +1,87 @@
+and the optional Composite phase which would come into play only when race conditions
+between component builds are detected.
+
+### Component phase
+
+When a single component image is built, the Integration Service tests the application
+by creating a Snapshot. All components from the Global Candidate List are included
+and the Component that was built is updated/overwritten with the newly built image.
+
+![](../diagrams/ADR-0015/component-phase.jpg)
+
+### Composite phase
+
+When multiple components are built at the same time, the Integration Service tests
+the application by creating a composite Snapshot using multiple components updated
+to use the newly built images.
+
+The Composite phase exists to resolve a race condition when teams merge multiple PRs
+to multiple components of the same application at nearly the same time.
+
+![](../diagrams/ADR-0015/composite-phase.jpg)
+
+## Consequences
+
+To illustrate the consequences of implementing the above approach,
+we can outline two scenarios, one with a happy path where only a single component
+is built at a time, and one with a race condition where two components are built
+in quick succession.
+
+### Happy path with a single component build
+
+In the happy path without race conditions, one PR merges to one component.
+1. The Component phase runs, and records results on the Snapshot.
+2. The Integration Service detects that the Snapshot is up-to-date by checking the
+Global Candidate List and skips the second phase of testing, marking the Snapshot as
+complete.
+3. The Snapshot is promoted according to the user preferences and the
+Global Candidate List is updated.
+
+### Path with a race condition
+
+In the path with race conditions, two PRs merge to two components at the same time.
+
+1. Two instances of the Component phase start up in parallel, testing two different
+Snapshots - each one including one of the two merged PRs.
+2. The first Snapshot finishes testing and detects that the Global Candidate List
+has not changed in the meantime and finalizes the Snapshot in question,
+updating the Global Candidate List with the newly built component image.
+3. The second Snapshot finishes testing. It is presented with an updated
+Global Candidate List that includes both changes, which have not been tested together yet.
+4. A Composite Snapshot is created that includes both changes and the Integration service
+starts testing it in the Composite phase.
+5. Composite phase testing runs in full and the Snapshot is promoted according to the
+user preferences
+
+## Footnotes
+For future consideration, if two components rely on each other to the point that
+breaking changes from one component create issues in another component during testing,
+a new feature to support batching of components together is being investigated.
+This would allow for the Integration service to hold off on testing a Component build
+until the dependent build is also completed.
+
```

Comment:
```suggestion

We originally made this decision verbally and with diagrams back in May of 2022, and have been operating with it as the operating design since then. However, we realized (through conversations like [this slack conversation](https://redhat-internal.slack.com/archives/C02CTEB3MMF/p1669961732100869)) that it is not obvious without documentation. We are documenting it here as an ADR for posterity, visibility. 
```

Reply:
Ack, added the proposed text below this line.